### PR TITLE
Rework internals to pass around `ares_dns_record_t` instead of binary data

### PR DESCRIPTION
**Pull request by bradh352**
c-ares has historically passed around raw dns packets in binary form.  Now that we have a new parser, and messages are already parsed internally, lets pass around that parsed message rather than requiring multiple parse attempts on the same message.  Also add a new `ares_send_dnsrec()` and `ares_query_dnsrec()` similar to `ares_search_dnsrec()` added with PR #719 that can return the pointer to the `ares_dns_record_t` to the caller enqueuing queries and rework `ares_search_dnsrec()` to use `ares_send_dnsrec()` internally.

### CODE REVIEW

**Comment by bradh352** [comment]
@oliverwelsh don't know if you want to review any progress, but I'm working on implementing what was discussed as the next step in #719

**Comment by bradh352** [comment]
Other than missing manpages, I think the code for this is done...

**Review by oliverwelsh — COMMENTED** [review]
Hi Brad, thank you for following up on this. It's great to see the DNS record object and parser used throughout the codebase as standard!

I have taken a look through of the code and left a couple of review comments. Overall I think this looks great. I especially like `ares__search_name_list()` - it feels a lot cleaner to calculate the search domains up front, which simplifies the logic when doing multiple searches.

File: src/lib/ares_query.c [hunk 1]
```diff
@@ -37,89 +37,147 @@
...
+  if (dnsrec != NULL) {
+    ares_status_t mystatus = ares_dns_write(dnsrec, &abuf, &alen);
+    if (mystatus != ARES_SUCCESS) {
+      status = mystatus;
     }
   }
+
+  qquery->callback(qquery->arg, (int)status, (int)timeouts, abuf, (int)alen);
+
   ares_free(qquery);
```
[the remark this answers is not in the document]
Comment (bradh352) [reply src/lib/ares_query.c:158]: ah, yep.  Wonder why the tests through ASAN didn't find that.  Thanks!

File: src/lib/ares_send.c [hunk 2]
```diff
@@ -150,18 +143,87 @@ ares_status_t ares_send_ex(ares_channel_t *channel, const unsigned char *qbuf,
...
+  size_t              alen  = 0;
+
+  if (dnsrec != NULL) {
+    status = ares_dns_write(dnsrec, &abuf, &alen);
+  }
+
+  cbarg->cb(cbarg->arg, (int)status, (int)timeouts, abuf, (int)alen);
+  ares_free(abuf);
+  ares_free(arg);
+}
```
[the remark this answers is not in the document]
Comment (bradh352) [reply src/lib/ares_send.c:187]: not a bad idea

File: src/lib/ares_dns_record.c [hunk 3]
```diff
@@ -1392,3 +1426,26 @@ ares_status_t ares_dns_record_create_query(ares_dns_record_t **dnsrec,
   }
   return status;
 }
+
+ares_dns_record_t *ares_dns_record_duplicate(const ares_dns_record_t *dnsrec)
```
[the remark this answers is not in the document]
Comment (bradh352) [reply src/lib/ares_dns_record.c:1430]: the ttl decrement is the only "internal", I put a note in the header about that limitation. Thanks.

File: src/lib/ares_dns_record.c [hunk 4]
```diff
@@ -276,6 +276,40 @@ ares_status_t ares_dns_record_query_add(ares_dns_record_t  *dnsrec,
   return ARES_SUCCESS;
 }
 
+ares_status_t ares_dns_record_query_set_name(ares_dns_record_t  *dnsrec,
```
[the remark this answers is not in the document]
Comment (bradh352) [reply src/lib/ares_dns_record.c:279]: Yes you are right, I'm pretty sure I'm not misusing it that way anywhere.  I'll add a note to the description letting people know this.

**Pull request merged**